File: Lite/src/main/org/xidea/el/parser/ExpressionTokenizer.java
package org.xidea.el.parser;

import static org.xidea.el.ExpressionToken.OP_ADD;
import static org.xidea.el.ExpressionToken.BIT_PARAM;
import static org.xidea.el.ExpressionToken.BIT_PRIORITY;
import static org.xidea.el.ExpressionToken.BIT_PRIORITY_SUB;
import static org.xidea.el.ExpressionToken.OP_AND;
import static org.xidea.el.ExpressionToken.OP_BIT_AND;
import static org.xidea.el.ExpressionToken.OP_BIT_OR;
import static org.xidea.el.ExpressionToken.OP_BIT_XOR;
import static org.xidea.el.ExpressionToken.OP_GET_PROP;

import static org.xidea.el.ExpressionToken.OP_MUL;
import static org.xidea.el.ExpressionToken.OP_DIV;
import static org.xidea.el.ExpressionToken.OP_MOD;
import static org.xidea.el.ExpressionToken.OP_LT;
import static org.xidea.el.ExpressionToken.OP_GT;
import static org.xidea.el.ExpressionToken.OP_LTEQ;
import static org.xidea.el.ExpressionToken.OP_GTEQ;
import static org.xidea.el.ExpressionToken.OP_EQ;
import static org.xidea.el.ExpressionToken.OP_NOTEQ;
import static org.xidea.el.ExpressionToken.OP_NOT;
import static org.xidea.el.ExpressionToken.OP_BIT_NOT;
import static org.xidea.el.ExpressionToken.OP_INVOKE_METHOD;
import static org.xidea.el.ExpressionToken.OP_MAP_PUSH;
import static org.xidea.el.ExpressionToken.OP_NEG;
import static org.xidea.el.ExpressionToken.OP_OR;
import static org.xidea.el.ExpressionToken.OP_PARAM_JOIN;
import static org.xidea.el.ExpressionToken.OP_POS;
import static org.xidea.el.ExpressionToken.OP_QUESTION;
import static org.xidea.el.ExpressionToken.OP_QUESTION_SELECT;
import static org.xidea.el.ExpressionToken.OP_GET_STATIC_PROP;
import static org.xidea.el.ExpressionToken.OP_SUB;
import static org.xidea.el.ExpressionToken.VALUE_CONSTANTS;
import static org.xidea.el.ExpressionToken.VALUE_NEW_LIST;
import static org.xidea.el.ExpressionToken.VALUE_NEW_MAP;
import static org.xidea.el.ExpressionToken.VALUE_VAR;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.xidea.el.ExpressionSyntaxException;
import org.xidea.el.ExpressionToken;
import org.xidea.el.json.JSONTokenizer;

public class ExpressionTokenizer extends JSONTokenizer {
	private final static TokenImpl TOKEN_TRUE = new TokenImpl(VALUE_CONSTANTS,
			Boolean.TRUE);
	private final static TokenImpl TOKEN_FALSE = new TokenImpl(VALUE_CONSTANTS,
			Boolean.FALSE);
	private final static TokenImpl TOKEN_NULL = new TokenImpl(VALUE_CONSTANTS,
			null);

	// 编译期间标记，compile time object
	static final int BRACKET_BEGIN = 0xFFFE;// ([{;
	static final int BRACKET_END = 0xFFFF;// )]};

	private static enum Status {
		STATUS_BEGIN, STATUS_EXPRESSION, STATUS_OPERATOR
	}

	private Status status = Status.STATUS_BEGIN;
	private int previousType = Integer.MIN_VALUE;

	protected ArrayList<TokenImpl> tokens = new ArrayList<TokenImpl>();
	protected TokenImpl expression;

	public ExpressionTokenizer(String value) {
		super(value);
		parseEL();
		prepareSelect();
		LinkedList<TokenImpl> stack = new LinkedList<TokenImpl>();
		try {
			toTree(right(this.tokens.iterator()), stack);
		} catch (Exception e) {
			throw new ExpressionSyntaxException(e);
		}
		if (stack.size() != 1) {
			new ExpressionSyntaxException("表达式语法错误：" + value);
		}
		this.expression = stack.getFirst();
	}

	private void prepareSelect() {
		int p1 = tokens.size();
		while (p1-- > 0) {
			int type1 = tokens.get(p1).getType();
			int dep = 0;
			int p2 = p1;
			int pos = 0;
			if (type1 == OP_QUESTION ) { //(a?b
				while(p2-->0) {
					int type2 = tokens.get(p2).getType();
					if (type2 > 0) {// op
						if (type2 == BRACKET_BEGIN) {
							dep++;
						} else if (type2 == BRACKET_END) {
							dep--;
						}
						if (dep == 0
								&& getPriority(type2) <= getPriority(OP_QUESTION)) {
							pos = p2+1;
//							System.out.println("&&&&&&");
							break;
						}
					}
				} 
				tokens.add(pos, new TokenImpl(BRACKET_BEGIN, null));
				p1++;
			}else if(type1 == OP_QUESTION_SELECT){
				int end = tokens.size();
				while(++p2<end) {
					int type2 = tokens.get(p2).getType();
					if (type2 > 0) {// op
						if (type2 == BRACKET_BEGIN) {
							dep++;
						} else if (type2 == BRACKET_END) {
							dep--;
						}
						if (dep == 0
								&& getPriority(type2) <= getPriority(OP_QUESTION)) {
							pos = p2;
//							System.out.println("###########");
							break;
						}
					}
				} 
				if(pos>0){
					tokens.add(pos, new TokenImpl(BRACKET_END, null));
				}else{
					tokens.add(new TokenImpl(BRACKET_END, null));
				}
			}
		}
	}

	public ExpressionToken getResult() {
		return expression;
	}

	private void toTree(Iterator<TokenImpl> tokens, LinkedList<TokenImpl> stack) {
		while (tokens.hasNext()) {
			final TokenImpl item = tokens.next();
			int type = item.getType();
			switch (type) {
			case VALUE_CONSTANTS:
			case VALUE_VAR:
			case VALUE_NEW_LIST:
			case VALUE_NEW_MAP:
				stack.addFirst(item);
				break;
			default:// OP
				if ((type & BIT_PARAM) > 0) {// 两个操作数
					TokenImpl arg2 = stack.removeFirst();
					TokenImpl arg1 = stack.removeFirst();
					item.setLeft(arg1);
					item.setRight(arg2);
					stack.addFirst(item);
				} else {// 一个操作树
					TokenImpl arg1 = stack.removeFirst();
					item.setLeft(arg1);
					stack.addFirst(item);
				}
			}
		}
	}

	// 将中序表达式转换为右序表达式
	private Iterator<TokenImpl> right(Iterator<TokenImpl> tokens) {
		LinkedList<List<TokenImpl>> rightStack = new LinkedList<List<TokenImpl>>();
		rightStack.addFirst(new ArrayList<TokenImpl>()); // 存储右序表达式

		LinkedList<TokenImpl> buffer = new LinkedList<TokenImpl>();

		while (tokens.hasNext()) {
			final TokenImpl item = tokens.next();
			if (item.getType() > 0) {
				if (buffer.isEmpty()) {
					buffer.addFirst(item);
				} else if (item.getType() == BRACKET_BEGIN) {// ("(")
					buffer.addFirst(item);
				} else if (item.getType() == BRACKET_END) {// .equals(")"))
					while (true) {
						TokenImpl operator = buffer.removeFirst();
						if (operator.getType() == BRACKET_BEGIN) {
							break;
						}
						addRightToken(rightStack, operator);
					}
				} else {
					while (!buffer.isEmpty()
							&& rightEnd(item, buffer.getFirst())) {
						TokenImpl operator = buffer.removeFirst();
						// if (operator.getType() !=
						// BRACKET_BEGIN){
						addRightToken(rightStack, operator);
					}
					buffer.addFirst(item);
				}
			} else {// lazy begin value exp
				addRightToken(rightStack, item);
			}
		}
		while (!buffer.isEmpty()) {
			TokenImpl operator = buffer.removeFirst();
			addRightToken(rightStack, operator);
		}
		return rightStack.getFirst().iterator();
	}

	private void addRightToken(LinkedList<List<TokenImpl>> rightStack,
			TokenImpl token) {
		List<TokenImpl> list = rightStack.getFirst();
		if (token.getType() == OP_GET_PROP) {
			int last = list.size() - 1;
			if (last >= 0) {
				TokenImpl previous = list.get(last);
				if (previous.getType() == VALUE_CONSTANTS) {
					list.remove(last);
					token = new TokenImpl(OP_GET_STATIC_PROP, previous
							.getParam());
				}
			}
		}
		list.add(token);
	}

	protected int getPriority(int type) {
		switch (type) {
		case BRACKET_BEGIN:
		case BRACKET_END:
			return Integer.MIN_VALUE;
		default:
			return (type & BIT_PRIORITY) << 4 | (type & BIT_PRIORITY_SUB) >> 12;
		}
	}

	private boolean rightEnd(TokenImpl item, TokenImpl privious) {
		int t1 = privious.getType();
		int t2 = item.getType();
		int p1 = getPriority(t1);
		int p2 = getPriority(t2);
		// 1+2*2
		// (a?b:c) == > (a?b):c
		// (a?b:ca:cb:cc) => (a?b):((ca?cb):cc)
		// 1?1:3 + 0?5:7 ==>1 //1?1:(3 + 0?5:7 )
		// 1?0?5:7:3 ==>7 //1?(0?5:7):3
		// 1?0?5:0?11:13:3 ==>13 //1?((0?5:0)?11:13):3
		if (p2 <= p1) {
			// if(p2 == p1){
			// if(t2 == OP_QUESTION_SELECT){
			// return true;//t1 == OP_QUESTION;
			// }else if(t2 == OP_QUESTION){
			// return false;//t1 == OP_QUESTION_SELECT;
			// }
			// }
			return true;
		} else {
			return false;
		}
	}

	protected void parseEL() {
		skipSpace(0);
		while (start < end) {
			char c = value.charAt(start);
			if (c == '"' || c == '\'') {
				String text = findString();
				addKeyOrObject(text, false);
			} else if (c >= '0' && c <= '9') {
				Number number = findNumber();
				addKeyOrObject(number, false);
			} else if (Character.isJavaIdentifierStart(c)) {
				String id = findId();
				if ("true".equals(id)) {
					addToken(TOKEN_TRUE);
				} else if ("false".equals(id)) {
					addToken(TOKEN_FALSE);
				} else if ("null".equals(id)) {
					addToken(TOKEN_NULL);
				} else {
					skipSpace(0);
					if (previousType == OP_GET_PROP) {
						addToken(new TokenImpl(VALUE_CONSTANTS, id));
					} else {
						addKeyOrObject(id, true);
					}
				}
			} else {
				String op = findOperator();
				// if (value.startsWith(op, start))
				if (op != null) {
					parseOperator(op);
				} else {
					this.reportError("未知操作符:");
				}
			}
			skipSpace(0);
		}
	}

	private String findOperator() {// optimize json ,:[{}]
		char c = value.charAt(start);
		int end = start + 1;
		char next = value.length() > end ? value.charAt(end) : 0;
		switch (c) {
		case ',':// optimize for json
		case ':':// 3op,map key
		case '[':// list
		case ']':
		case '{':// map
		case '}':
		case '(':// quote
		case ')':
		case '.':// prop
		case '?':// 3op
		case '+':// 5op
		case '-':
		case '~':
		case '^':
		case '*':
		case '/':
		case '%':
			break;
		case '=':// ==
			if (next == '=') {
				end++;
				if (value.length() > end && value.charAt(end) == '=') {
					this.reportError("不支持=== 和!==操作符，请使用==,!=");
				}
			} else {
				this.reportError("不支持赋值操作:");
			}
			break;
		case '!':// !,!=
			if (next == '=') {
				end++;
				if (value.length() > end && value.charAt(end) == '=') {
					this.reportError("不支持=== 和!==操作符，请使用==,!=");
				}
			}
			break;
		case '>':// >,>=
		case '<':// <,<=
			if (next == '=') {
				end++;
			}
			break;
		case '&':// && / &
		case '|':// || /|
			if ((c == next)) {
				end++;
			}
			break;
		default:
			return null;
		}

		return value.substring(start, start = end);
	}

	private void reportError(String msg) {
		throw new ExpressionSyntaxException(msg + "\n@" + start + "\n"
				+ value.substring(start) + "\n----\n" + value);
	}

	/**
	 * 碰見:和,的時候，就需要檢查是否事map的間隔符號了
	 * 
	 * @return
	 */
	private boolean isMapMethod() {
		int i = tokens.size() - 1;
		int depth = 0;
		for (; i >= 0; i--) {
			TokenImpl token = tokens.get(i);
			int type = token.getType();
			if (depth == 0) {
				if (type == OP_MAP_PUSH || type == VALUE_NEW_MAP) {// (
					// <#newMap>
					// <#push>
					return true;
				} else if (type == OP_PARAM_JOIN) {// (
					// <#newList>
					// <#param_join>
					return false;
				}
			}
			if (type == BRACKET_BEGIN) {
				depth--;
			} else if (type == BRACKET_END) {
				depth++;
			}
		}
		return false;
	}

	private void parseOperator(String op) {
		if (op.length() == 1) {
			switch (op.charAt(0)) {
			case '(':
				if (status == Status.STATUS_EXPRESSION) {
					addToken(new TokenImpl(OP_INVOKE_METHOD, null));
					if (skipSpace(')')) {
						addToken(new TokenImpl(VALUE_CONSTANTS,
								Collections.EMPTY_LIST));
						start++;
					} else {
						addList();
					}

				} else {
					addToken(new TokenImpl(BRACKET_BEGIN, null));
				}
				break;
			case '[':
				if (status == Status.STATUS_EXPRESSION) {// getProperty
					addToken(new TokenImpl(OP_GET_PROP, null));
					addToken(new TokenImpl(BRACKET_BEGIN, null));
				} else {// list
					addList();
				}
				break;
			case '{':
				addMap();
				break;
			case '}':
			case ']':
			case ')':
				addToken(new TokenImpl(BRACKET_END, null));
				break;
			case '+'://
				addToken(new TokenImpl(
						status == Status.STATUS_EXPRESSION ? OP_ADD : OP_POS,
						null));
				break;
			case '-':
				addToken(new TokenImpl(
						status == Status.STATUS_EXPRESSION ? OP_SUB : OP_NEG,
						null));
				break;
			case ',':// :(object_setter is skiped,',' should
				// be skip)
				if (!isMapMethod()) {
					addToken(new TokenImpl(OP_PARAM_JOIN, null));

				}
				break;
			case '/':
				char next = value.charAt(start);
				if (next == '/' || next == '*') {
					start--;
					skipComment();
					break;
				} else if (this.status != Status.STATUS_EXPRESSION) {
					int end = findRegExp(this.value, this.start);
					if (end > 0) {
						String regexp = this.value.substring(this.start - 1,
								end);
						HashMap<String, String> value = new HashMap<String, String>();
						value.put("class", "RegExp");
						value.put("source", regexp);
						this.addToken(new TokenImpl(VALUE_CONSTANTS, value));
						this.start = end;
						break;
					}
				}
				addToken(new TokenImpl(TOKEN_MAP.get(op), null));
				break;
			default:
				addToken(new TokenImpl(TOKEN_MAP.get(op), null));
			}
		} else {
			addToken(new TokenImpl(TOKEN_MAP.get(op), null));
		}

	}

	private void addToken(TokenImpl token) {
		switch (token.getType()) {
		case BRACKET_BEGIN:
			status = Status.STATUS_BEGIN;
			break;
		case VALUE_CONSTANTS:
		case VALUE_VAR:
		case BRACKET_END:
			status = Status.STATUS_EXPRESSION;
			break;
		default:
			status = Status.STATUS_OPERATOR;
			break;
		}
		// previousType2 = previousType;
		previousType = token.getType();
		tokens.add(token);
	}

	private void addKeyOrObject(Object object, boolean isVar) {
		if (skipSpace(':') && isMapMethod()) {// object key
			addToken(new TokenImpl(OP_MAP_PUSH, object));
			this.start++;// skip :
		} else if (isVar) {
			addToken(new TokenImpl(VALUE_VAR, object));
		} else {
			addToken(new TokenImpl(VALUE_CONSTANTS, object));
		}
	}

	private void addList() {
		addToken(new TokenImpl(BRACKET_BEGIN, null));
		addToken(new TokenImpl(VALUE_NEW_LIST, null));
		if (!skipSpace(']')) {
			addToken(new TokenImpl(OP_PARAM_JOIN, null));
		}
	}

	private void addMap() {
		addToken(new TokenImpl(BRACKET_BEGIN, null));
		addToken(new TokenImpl(VALUE_NEW_MAP, null));
	}

	int findRegExp(String text, int start) {
		int depth = 0;
		int end = text.length();
		char c;
		while (start < end) {
			c = text.charAt(start++);
			if (c == '[') {
				depth = 1;
			} else if (c == ']') {
				depth = 0;
			} else if (c == '\\') {
				start++;
			} else if (depth == 0 && c == '/') {
				while (start < end) {
					c = text.charAt(start++);
					switch (c) {
					case 'g':
					case 'i':
					case 'm':
						break;
					default:
						return start - 1;
					}
				}

			}
		}
		return -1;
	}

	private static final Map<String, Integer> TOKEN_MAP = new HashMap<String, Integer>();
	static {

		// 9
		TOKEN_MAP.put(".", OP_GET_PROP);
		// 8
		TOKEN_MAP.put("!", OP_NOT);
		TOKEN_MAP.put("^", OP_BIT_NOT);
		// TOKEN_MAP.put("+",OP_POS);
		// TOKEN_MAP.put("-",OP_NEG);
		// 7
		TOKEN_MAP.put("*", OP_MUL);
		TOKEN_MAP.put("/", OP_DIV);
		TOKEN_MAP.put("%", OP_MOD);
		// 6
		// TOKEN_MAP.put("+",OP_ADD);
		// TOKEN_MAP.put("-",OP_SUB);
		// 5
		TOKEN_MAP.put("<", OP_LT);
		TOKEN_MAP.put(">", OP_GT);
		TOKEN_MAP.put("<=", OP_LTEQ);
		TOKEN_MAP.put(">=", OP_GTEQ);
		TOKEN_MAP.put("==", OP_EQ);
		TOKEN_MAP.put("!=", OP_NOTEQ);

		// 4
		TOKEN_MAP.put("&", OP_BIT_AND);
		TOKEN_MAP.put("^", OP_BIT_XOR);
		TOKEN_MAP.put("|", OP_BIT_OR);
		// 3
		TOKEN_MAP.put("&&", OP_AND);
		TOKEN_MAP.put("||", OP_OR);
		// 2
		TOKEN_MAP.put("?", OP_QUESTION);
		TOKEN_MAP.put(":", OP_QUESTION_SELECT);// map 中的：被直接skip了
		// 1
		TOKEN_MAP.put(",", OP_PARAM_JOIN);

		// OP_MAP_PUSH
		// OP_INVOKE_METHOD
	}

}

File: trunk/Lite/web/scripts/org/xidea/lite/expression-tokenizer.js
var STATUS_BEGIN = -100;
var STATUS_EXPRESSION = -101;
var STATUS_OPERATOR = -102;
var fns = {
	toTokens :function() {
		return this.expression;//.slice(0).reverse();// reversed
	},

	parseEL :function() {
		this.skipSpace(0);
		while (this.start < this.end) {
			var c = this.value.charAt(this.start);
			if (c == '"' || c == '\'') {
				var text = this.findString();
				this.addKeyOrObject(text, false);
			} else if (c >= '0' && c <= '9') {
				var number = this.findNumber();
				this.addKeyOrObject(number, false);
			} else if (/[\w$_]/.test(c)) {
				var id = this.findId();
				switch(id){
				    case 'true':
				        this.addToken([VALUE_CONSTANTS,true]);
				        break;
				    case 'false':
				        this.addToken([VALUE_CONSTANTS,false]);
				        break;
				    case 'null':
				        this.addToken([VALUE_CONSTANTS,null]);
				        break;
				    default:
    				    this.skipSpace(0);
    					if (this.previousType == OP_GET_PROP) {
    						this.addToken([VALUE_CONSTANTS,
    								id]);
    					} else {
    						this.addKeyOrObject(id, true);
    					}
				}
			} else {
				var op = this.findOperator();
				// if (this.value.startsWith(op, this.start))
				this.parseOperator(op);
				if (op == null) {
					throw new Error("语法错误:" + this.value + "@"
							+ this.start);
				}
			}
			this.skipSpace(0);
		}
	},

	findOperator :function() {// optimize json ,:[{}]
		switch (this.value.charAt(this.start)) {
		case '!':// !,!=
		case '>':// >,>=
		case '<':// <,<=
			if (this.value.charAt(this.start + 1) == '=') {
				return this.value.substring(this.start, this.start += 2);
			}
		case ',':// optimize for json
		case ':':// 3op,map key
		case '[':// list
		case ']':
		case '{':// map
		case '}':
		case '(':// quote
		case ')':
		case '.':// prop
		case '?':// 3op
		case '+':// 5op
		case '-':
		case '*':
		case '/':
		case '%':
			return this.value.substring(this.start, this.start += 1);

		case '=':// ==
		case '&':// &&
		case '|':// ||
    		if(":debug"){
    			if (this.value.charAt(this.start) != this.value.charAt(this.start + 1)){
    			    $log.error("Error OP:" + this.value.substring(this.start,this.start+2))
    			}
    		}
			return this.value.substring(this.start, this.start += 2);
		}
		alert(this.value.substring(this.start))
		return null;
	},

	/**
	 * 碰見:和,的時候，就需要檢查是否事map的間隔符號了
	 * 
	 * @return
	 */
	isMapMethod :function() {
		var i = this.tokens.length - 1;
		var depth = 0;
		for (; i >= 0; i--) {
			var token = this.tokens[i];
			var type = token[0];
			if (depth == 0) {
				if (type == OP_MAP_PUSH
						|| type == VALUE_NEW_MAP) {// (
					// <#newMap>
					// <#push>
					return true;
				} else if (type == OP_PARAM_JOIN) {// (
					// <#newList>
					// <#param_join>
					return false;
				}
			}
			if (type == BRACKET_BEGIN) {
				depth--;
			} else if (type == BRACKET_END) {
				depth++;
			}
		}
		return false;
	},

	parseOperator :function(op) {
		if (op.length == 1) {
			switch (op.charAt(0)) {
			case '(':
				if (this.status == STATUS_EXPRESSION) {
					this.addToken([OP_INVOKE_METHOD, null]);
					if (this.skipSpace(')')) {
						this.addToken([VALUE_CONSTANTS,
								[]]);
						this.start++;
					} else {
						this.addList();
					}

				} else {
					this.addToken([BRACKET_BEGIN, null]);
				}
				break;
			case '[':
				if (this.status == STATUS_BEGIN || this.status == STATUS_OPERATOR) {// list
					this.addList();

				} else if (this.status == STATUS_EXPRESSION) {// getProperty
					this.addToken([OP_GET_PROP, null]);
					this.addToken([BRACKET_BEGIN, null]);
				} else {
					throw new Error("语法错误:" + this.value + "@"
							+ this.start);
				}
				break;
			case '{':
				this.addMap();
				break;
			case '}':
			case ']':
			case ')':
				this.addToken([BRACKET_END, null]);
				break;
			case '+'://
				this.addToken([
						this.status == STATUS_OPERATOR ? OP_POS
								: OP_ADD, null]);
				// this.addToken(OperatorToken.getToken(SKIP_AND));
				break;
			case '-':
				this.addToken([
						this.status == STATUS_OPERATOR ? OP_NEG
								: OP_SUB, null]);
				// this.addToken(OperatorToken.getToken(SKIP_AND));
				break;
			case '?':// ?:
				this.addToken([OP_QUESTION, null]);
				// this.addToken(OperatorToken.getToken(SKIP_QUESTION));
				this.addToken([VALUE_LAZY, null]);
				break;
			case ':':// :(object_setter is skiped)
				this.addToken([OP_QUESTION_SELECT, null]);
				this.addToken([VALUE_LAZY, null]);
				break;
			case ',':// :(object_setter is skiped,',' should
				// be skip)
				if (!this.isMapMethod()) {
					this.addToken([OP_PARAM_JOIN, null]);

				}
				break;
			case '/':
				var next = this.value.charAt(this.start);
				if (next == '/') {
					var end1 = this.value.indexOf('\n', this.start);
					var end2 = this.value.indexOf('\r', this.start);
					var cend = Math.min(end1, end2);
					if (cend < 0) {
						cend = Math.max(end1, end2);
					}
					if (cend > 0) {
						this.start = cend;
					} else {
						this.start = this.end;
					}
					break;
				} else if (next == '*') {
					var cend = this.value.indexOf("*/", this.start);
					if (cend > 0) {
						this.start = cend + 2;
					} else {
						throw new Error("未結束注釋:" + this.value
								+ "@" + this.start);
					}
					break;
				}
			default:
				this.addToken([findTokenType(op), null]);
			}
		} else if (op == "||") { // ||
			this.addToken([OP_OR, null]);
			this.addToken([VALUE_LAZY, null]);
			// this.addToken(LazyToken.LAZY_TOKEN_END);
		} else if (op == "&&") {// &&
			this.addToken([OP_AND, null]);
			this.addToken([VALUE_LAZY, null]);
			// this.addToken(OperatorToken.getToken(SKIP_AND));
		} else {
			this.addToken([findTokenType(op), null]);
		}

	},

	addToken :function(token) {
		switch (token[0]) {
		case BRACKET_BEGIN:
			this.status = STATUS_BEGIN;
			break;
		case VALUE_CONSTANTS:
		case VALUE_VAR:
		case BRACKET_END:
			this.status = STATUS_EXPRESSION;
			break;
		default:
			this.status = STATUS_OPERATOR;
			break;
		}
		// previousType2 = this.previousType;
		this.previousType = token[0];
		this.tokens.push(token);
	},

	addKeyOrObject :function(object, isVar) {
		if (this.skipSpace(':') && this.isMapMethod()) {// object key
			this.addToken([OP_MAP_PUSH, object]);
			this.start++;// skip :
		} else if (isVar) {
			this.addToken([VALUE_VAR, object]);
		} else {
			this.addToken([VALUE_CONSTANTS, object]);
		}
	},

	addList :function() {
		this.addToken([BRACKET_BEGIN, null]);
		this.addToken([VALUE_NEW_LIST, null]);
		if (!this.skipSpace(']')) {
			this.addToken([OP_PARAM_JOIN, null]);
		}
	},

	addMap :function() {
		this.addToken([BRACKET_BEGIN, null]);
		this.addToken([VALUE_NEW_MAP, null]);
	}
};
var pt = new JSONTokenizer('');
for(var n in fns){
    pt[n] = fns[n]
}
function ExpressionTokenizer(value){
    this.value = value.replace(/^\s+|\s+$/g,'');
	this.start = 0;
	this.end = this.value.length;
    this.status = STATUS_BEGIN;
	this.previousType = STATUS_BEGIN;
	this.tokens = [];
	this.parseEL();
	this.expression = trimToken(right(this.tokens));
	
}
ExpressionTokenizer.prototype = pt;



// 将中序表达式转换为右序表达式
function right(tokens) {
	var rightStack = [[]];
	var buffer = [];

	for (var i = 0;i<tokens.length;i++) {
		var item = tokens[i];
		if (item[0] > 3) {
			if (buffer.length == 0) {
				buffer.push(item);
			} else if (item[0] == BRACKET_BEGIN) {// ("(")
				buffer.push(item);
			} else if (item[0] == BRACKET_END) {// .equals(")"))
				while (true) {
					var operator = buffer.pop();
					if (operator[0] == BRACKET_BEGIN) {
						break;
					}
					addRightOperator(rightStack, operator);
				}
			} else {
				while (buffer.length!=0
						&& rightEnd(item[0], buffer[buffer.length-1][0])) {
					var operator = buffer.pop();
					// if (operator[0] !=
					// BRACKET_BEGIN){
					addRightOperator(rightStack, operator);
				}
				buffer.push(item);
			}
		} else {// lazy begin value exp
			addRightToken(rightStack, item);
		}
	}
	while (buffer.length !=0) {
		var operator = buffer.pop();
		addRightOperator(rightStack, operator);
	}
	return rightStack[rightStack.length-1];
}
function trimToken(tokens){
	for(var i=0;i<tokens.length;i++){
		var token = tokens[i];
		token.length = getTokenLength(token[0]);
	}
	return tokens;
}
function addRightOperator(rightStack,
		operator) {
	switch (operator[0]) {
	case OP_OR:
	case OP_AND:
	case OP_QUESTION:
	case OP_QUESTION_SELECT:
		var children = rightStack.pop();
		var list = rightStack[rightStack.length-1];
		if (children.length == 1) {
			list[list.length - 1]= children[0];
		} else {
			var token = list[list.length - 1];
			token[1]=trimToken(children);//.slice(0).reverse();;

		}
	}
	addRightToken(rightStack, operator);
}

function addRightToken(rightStack,
		token) {
	var list = rightStack[rightStack.length-1];
	if (token[0] == OP_GET_PROP) {
	    var last = list.length-1;
	    if(last>=0){
	        var previous = list[last];
	        if(previous[0] == VALUE_CONSTANTS){
	            list.length--;
	            token = [OP_STATIC_GET_PROP,previous[1]]; 
	        }
	    }
	}else if (token[0] == VALUE_LAZY) {
		rightStack.push([]);
	}
	token.toString = toTokenString;
	list.push(token);
}

function getPriority(type) {
	switch (type) {
	case BRACKET_BEGIN:
	case BRACKET_END:
		return Math.MIN_VALUE;
	case OP_GET_PROP:
	//case OP_GET_METHOD:
	//case OP_GET_GLOBAL_METHOD:
	case OP_INVOKE_METHOD:
	case VALUE_NEW_LIST:
	case VALUE_NEW_MAP:
		return 12;

	case OP_NOT:
	case OP_POS:
	case OP_NEG:
		return 8;

	case OP_MUL:
	case OP_DIV:
	case OP_MOD:
		return 4;

	case OP_ADD:
	case OP_SUB:
		return 1;

	case OP_LT:
	case OP_GT:
	case OP_LTEQ:
	case OP_GTEQ:
	case OP_EQ:
	case OP_NOTEQ:
		return 0;

	case OP_AND:
		return -1;
	case OP_OR:
		return -2;

	case OP_QUESTION:
	case OP_QUESTION_SELECT:
		return -4;// !!

	case OP_MAP_PUSH:
		return -7;// !!
	case OP_PARAM_JOIN:
		return -8;
	}
	throw new Error("unsupport token:" + type);
}

function rightEnd(itemType, priviousType) {
	return getPriority(itemType) <= getPriority( priviousType);
}